torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1724,16 +1765,38 @@ struct to_ir {
     };
     GatheredTypes gathered(typeParser_);
     gathered.gather(classinfo);
-    auto val = emitExpr(obj);
+    Value* val = emitExpr(obj);
     RefinementSet refinement;
     if (gathered.types.size() == 1 &&
         gathered.types.at(0)->isSubtypeOf(val->type()) &&
         obj.kind() == TK_VAR) {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+      Refinement isinstance(ident, gathered.types.at(0));
```

Comment:
Can you factor out the refinement logic into a separate stacked PR? I am having a really hard time parsing through the logic and the 500 lines to verify that this is properly being tested.  There was the previous bug I commented with refining Unions even though @gmagogsfm which is a good indication that this PR is too large for us to be confident about code quality